0015-authority.md

```diff
@@ -0,0 +1,501 @@
+        }
+        throw new SpawnStateException($"Cannot get authority for not spawned {nameof(NetworkObject)}");
+    }
+}
+```
+
+There will also be a function to change authority. Similar to ownership this will use the `NetworkSpawnManager` to send authority changes to all clients.
+```csharp
+public void ChangeAuthority(Authority authority)
+{
+//.....
+}
+```
+
+## NetworkVariable Authority
+
+We introduce a `NetworkVariablePermission.Authority` permission. This is a very simple change to the NetworkVariable we just add the following to the read/write queries.
+
+This permission is needed so that authority is not just a gameplay code feature but can also be enforced onto NetworkVariables so that server authoritative games can remain authoritative
+
+```csharp
+public bool CanClientRead(ulong clientId)
+{
+    switch (Settings.ReadPermission)
+    {
+    //....
+        case NetworkVariablePermission.Authority:
+            return m_NetworkBehaviour.HasAuthority;
+    //.....
+    }
+    return true;
+}
+```
+
+## OnGainedAuthority / OnLostAuthority
+
+Will be implemented by listening to authority mode changes internally and then checking whether the resulting `HasAuthority` flag is different.
+
+These have similar usability as `OnGainedOwnership/OnLostOwnership`, in most cases they are not necessary but can be helpful.
+
+## SupportedAuthorityModes Attribute
+
+We will introduce a supported authority modes attribute. The main purpose for this attribute is to signal compatibility with other components.
+
+The attribute is optional, if no attribute is added to a component all authority modes are considered to be supported. The attribute will not strongly enforce anything but just display warnings in the console and on NetworkObject if a mode is not supported.
+
+```csharp
+    [AttributeUsage(AttributeTargets.Class)]
+    public class SupportedAuthorityModesAttribute : Attribute
+    {
+        private Authority m_AuthorityModes;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool SupportsAuthority(SupportedAuthorityModesAttribute attribute, Authority authority)
+        {
+            // Enum.HasFlag but faster
+            return (((byte)attribute.m_AuthorityModes & (int)authority) != 0);
+        }
+
+        public SupportedAuthorityModesAttribute(Authority authorityModes)
+        {
+            m_AuthorityModes = authorityModes;
+        }
+    }
+```
+
+At runtime modes get cached in a static lookup table `Dictionary<Type, Authority>` to avoid reflection in the hot path.
+
+Supported Authority mode just changes some minor details in NetworkObjectEditor to display components with not supported authority and after changing authority it will check whether the behaviour is still supported. In addition it checks foreach NetworkBehaviour whether the authority mode is still allowed when authority gets changed and if not logs a warning to the console.
+
+# Drawbacks
```

Comment:
I think it's worth calling out in the drawbacks that inexperienced developers might be confused by the difference between ownership and authority. I don't think that's a reason not to implement authority, but it is definitely a drawback to making that separation and will need to be well-documented.